### PR TITLE
Attribute editor & OSM importer

### PR DESCRIPTION
**Pull request by pedrocamargo**
A LOT of changes here that go beyond the attribute editor and the OSM importer

### CODE REVIEW

**Comment by pedrocamargo** [comment]
I had opened the pull request against master instead of the new version development branch!

File: aequilibrae/parameters.py [hunk 1]
```diff
@@ -11,24 +11,25 @@ class Parameters:
     Parameters are used in many procedures, and are often defined only in thi parameters.yml file ONLY
```
Comment (jamiecook) [inline aequilibrae/parameters.py:11]:
```suggestion
    Parameters are used in many procedures, and are often defined only in the parameters.yml file ONLY
```

File: aequilibrae/project/database_specification/triggers/network_triggers.sql [hunk 2]
```diff
@@ -356,37 +356,37 @@ BEGIN
...
+--    FROM nodes
+--    WHERE node_id == new.b_node
+--    AND geometry = EndPoint(new.geometry) AND
+--    ROWID IN (
+--      SELECT ROWID FROM SpatialIndex WHERE f_table_name = 'nodes' AND
+--      search_frame = new.geometry)) == 0
+--
+--BEGIN
+--	SELECT raise(ABORT, 'This change to the b_node field is wrong.');
+--END;
```
Comment (jamiecook) [inline aequilibrae/project/database_specification/triggers/network_triggers.sql:392]: Is this gone now?

File: aequilibrae/project/field_editor.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        return dt
+
+    def __run_query_commit(self, qry: str, values=None) -> None:
+        conn = database_connection()
+        if values is None:
+            conn.execute(qry)
+        else:
+            conn.execute(qry, values)
+        conn.commit()
+        conn.close()
```
Comment (jamiecook) [inline aequilibrae/project/field_editor.py:155]: Feel like this should be in some utility class

File: aequilibrae/project/database_specification/triggers/network_triggers.sql (hunk 2 above)
Comment (pedrocamargo) [reply aequilibrae/project/database_specification/triggers/network_triggers.sql:392]: I am trying to get around some issues to get it working, but gone for now

File: aequilibrae/project/field_editor.py (hunk 3 above)
Comment (pedrocamargo) [reply aequilibrae/project/field_editor.py:155]: What do you mean?

File: aequilibrae/project/database_specification/triggers/network_triggers.sql (hunk 2 above)
Comment (jamiecook) [reply aequilibrae/project/database_specification/triggers/network_triggers.sql:392]: remove it?

File: aequilibrae/project/network/network.py [hunk 4]
```diff
@@ -173,24 +199,37 @@ def create_from_osm(
...
-        ignore_fields = ['ogc_fid', 'geometry']
-        all_fields = [f[1] for f in field_names if f[1] not in ignore_fields]
+        if fields is None:
+            curr.execute('PRAGMA table_info(links);')
+            field_names = curr.fetchall()
+
+            ignore_fields = ['ogc_fid', 'geometry']
+            all_fields = [f[1] for f in field_names if f[1] not in ignore_fields]
+        else:
+            all_fields = fields
```
Comment (jamiecook) [inline aequilibrae/project/network/network.py:232]:
```suggestion
        if fields is None:
            curr.execute('PRAGMA table_info(links);')
            field_names = curr.fetchall()

            ignore_fields = ['ogc_fid', 'geometry']
            fields = [f[1] for f in field_names if f[1] not in ignore_fields]
```

and then just use fields in the rest of the method?

File: aequilibrae/project/network/osm_builder.py [hunk 5]
```diff
@@ -46,37 +52,28 @@ def doWork(self):
...
-            self.__emit_all(["Value", i])
+        tot_items = len(self.osm_items)
+        for i in range(tot_items, 0, -1):
+            item = self.osm_items.pop(-1)
+            if item['type'] == "way":
+                alinks.append(item)
+            elif item['type'] == "node":
+                n.append(item)
+            self.__emit_all(["Value", tot_items - i])
+        gc.collect()
```
Comment (jamiecook) [inline aequilibrae/project/network/osm_builder.py:75]: This is modifying osm_items as it loops through them?
Are they memory hungry?
I'd suggest putting some comments around this loop

File: aequilibrae/project/network/osm_builder.py [hunk 6]
```diff
@@ -115,22 +112,25 @@ def importing_links(self, node_count):
...
 
-        nodes_to_add = set()
         counter = 0
         mode_codes, not_found_tags = self.modes_per_link_type()
+        owf, twf = self.field_osm_source()
 
         for osm_id, link in self.links.items():
             self.__emit_all(["Value", counter])
             counter += 1
+            if divmod(counter, 1000)[1] == 0:
```
Comment (jamiecook) [inline aequilibrae/project/network/osm_builder.py:130]:
```suggestion
            if counter % 1000 == 0:
```

File: aequilibrae/project/network/osm_builder.py [hunk 7]
```diff
@@ -144,76 +144,67 @@ def importing_links(self, node_count):
...
+
+    def __update_table_structure(self):
+        curr = self.conn.cursor()
+        curr.execute('pragma table_info(Links)')
+        structure = curr.fetchall()
+        has_fields = [x[1].lower() for x in structure]
+        fields = [field.lower() for field in self.get_link_fields()] + ['osm_id']
+        for field in fields:
+            if field in has_fields:
+                continue
```
Comment (jamiecook) [inline aequilibrae/project/network/osm_builder.py:204]:
```suggestion
        for field in [f for f in fields if not f in has_fields]:
```

**Review by jamiecook — APPROVED** [review]

File: aequilibrae/project/database_specification/triggers/network_triggers.sql (hunk 2 above)
Comment (pedrocamargo) [reply aequilibrae/project/database_specification/triggers/network_triggers.sql:392]: I should...

**Comment by pedrocamargo** [comment]
Thanks, @jamiecook !!

**Pull request merged**